New Actual Rates model : set rates in deg/s, independently of each other

ctzsnooze opened the pull request:
This PR proposes the introduction of a new method for configuring Rates, based on ideas I've had for quite a long time.  Big THANK YOU to IllusionFPV for providing me with the impetus to actually do something, and for his coding assistance.

The classic Betaflight rates model uses three 8 bit parameters:

- rcRrate, which returns a linear relationship between stick position and rate, such that the maximum rate in deg/s, at full stick, is 200 times the Configurator value (twice the CLI number).

- SRate, which exponentially increases the preceding linear rate as the sticks move out from centre. It has little or no effect at centre position. The inflexion point of the curve cannot be changed.  Adding more SRate will make for a steeper looking curve because the maximum rate increases a lot at maximum stick deflection.  An SRate value of 0.5 in Configurator (50 in CLI) will double the rate at maximum stick position.

- Expo, which adds a second 'curve' on top of the preceding values, with greater effect around centre stick position, and no effect on maximum rate.  The inflexion point of the curve is a bit different from the SRate curve, and changes with the amount of expo.  An expo value of 0.5 in Configurator will flatten the rate curve around centre stick by a factor of 2x, but will not change maximum rate.

When we configure our rates, we primarily care about three things:

- centre stick 'sensitivity', or, how 'twitchy' is the centre stick responsiveness?  Technically, this is determined by the rate change of rate per unit stick deflection around centre position.  

- the maximum roll rate at full stick deflection, i.e., how fast will it spin when we do a flip.

- how quickly the responsiveness changes as we move the stick from centre to full deflection, or 'expo inflexion point', meaning, whether the change in responsiveness increases steadily and smoothly as the sticks move further out (low expo, as used by racers), or whether we get a wide range of relatively low responsiveness and then a sudden increase at max rates (high expo, as often used in freestyle).  

While Betaflight rates have been around a long time, the method has some drawbacks:

- cannot set centre sensitivity or max rate in degrees/s
- each parameter interacts with another in terms of its effect
- limited control over the expo transition point
- weird config numbers that don't directly relate to outcomes
- config numbers that are different in CLI and Configurator

This summarises the interactions between Betaflight rate settings and outcomes:

- centre 'sensitivity' is affected by both rcRate and expo
- max roll rate is affected by both rcRate and SRate
- expo inflexion point is affected by both SRate and expo

This PR proposes the eventual replacement of the current Betaflight rates configuration paradigm with a different method.  The goals are:

- let the user set centre sensitivity, max roll rate and expo inflexion directly
- ensure that value is independent of any other
- set centre sensitivity and max roll rate in deg/s
- ultimately, show the same number in CLI and Configurator (requires 16bit values)

The way the code works is:

- first, use the value for 'centre sensitivity' to directly configure a 'baseline' linear response to stick movement.  The value entered for 'centre sensitivity' represents the actual stick sensitivity at centre in degrees/sec. The number entered is the centre 'feel' you'd get for a linear rate configured at the set deg/s.  This is the kind of 'base' responsiveness.

- let the user set their desired max turn rate in deg/s, and make a 'wedge shaped' linear ramp of the extra rate amount, above baseline, that is needed to get to that max rate at full deflection.

- apply an expo factor that changes the inflexion point of that ramped quantity from nearly linear to rising sharply at extreme stick deflection.  

Ideally, the centre sensitivity and max rate would be entered directly as deg/s.  For example, a freestyle setup might be centre sensitivity of 120 deg/s and max rate of 800 deg/s, and a race setup might range from 150 to 500.  In this PR, I have to re-use the rcRate and Rate 8 bit values.

To enter values as per this PR, we have to use 8 bit numbers, as follows:

- Centre sensitivity: for 150, set rcRate to 0.15 in Configurator,  or 15 in CLI.
- Max Rate: for 800, set SRate to 0.8 in Configurator, or 80 in CLI.
- Expo: set the same as before.  0.5 in Configurator, or 50 in CLI.

To determine your current centre sensitivity from the classic Betaflight rates:

- if you have no expo, it is rcRate * 200
- if you ave expo, it is rcRate * 200, attenuatedby (1-expo). 

So if you had an rcRate of 0.75 and no expo, your current centre sensitivity is 150.  If you have rcRate of 1.0, and 0.5 of expo, your current centre sensitivity is 100.

Your max rate is shown by the Rates curve in Configurator.

Once you know your current centre sensitivity and max rate, you can switch to the new `ACTUAL` rates method, and enter these values directly.

The new expo factor is a true inflexion point curve adjustment, and can't be directly translated from any previous Betaflight settings since they have complex interactions.  A starting value of 0.5 is fairly typical.  When set to 0, the stick responsiveness changes more smoothly than is possible currently.  When set to 1, the expo is quite extreme.  

Note that changing expo will not change your centre sensitivity or your max rate.  Centre feel will differ with low expo because we get a greater increase in rates around centre than with higher expo, but the actual dead centre sensititity to fine inputs will not change.

ctzsnooze commented:
Here are some graphics showing how old and new values can translate one to the other... it is possible to recreate all except the most extreme expo configurations in the new model.

First, my normal race rates, showing how expo 0.5 on the new setup allows me to recreate what I had...

![Screen Shot 2020-02-19 at 09 42 39](https://user-images.githubusercontent.com/11737748/74784557-4882f680-52fc-11ea-8c14-1c615b7d03f3.jpg)

Now a freestyle setup with quite extreme expo...

![Screen Shot 2020-02-19 at 09 46 46](https://user-images.githubusercontent.com/11737748/74784775-cb0bb600-52fc-11ea-8f94-fd565649d4a2.jpg)

ctzsnooze commented:
These images show how changing the new expo value changes the nature of the inflexion point, without changing the sensitivity at either end:

First, expo of zero, with centre sensitivity 160 and max rate 800, we get less expo than is possible currently...
![Screen Shot 2020-02-19 at 09 50 24](https://user-images.githubusercontent.com/11737748/74785033-6bfa7100-52fd-11ea-9f9a-6af52b69c493.jpg)

Now with expo at 0.7 we get something very similar to default expo from SRate when configured at that rate.  

![Screen Shot 2020-02-19 at 09 52 33](https://user-images.githubusercontent.com/11737748/74785113-a106c380-52fd-11ea-8599-4856098a1e9a.jpg)

With expo at 1, we get a steeper expo curve than currently from SRate alone.  Note two things:
- First, how at higher expo values, the rate line is fairly flat initially.  This gives equalled stick sensitivity around the centre.  
- Second, that the expo factor is only changing where the transition point takes off.  The transition will be predictable for any specified expo value, and is not affected by the user's centre or max rate values.  Only the single expo setting is required (not two, at present) for a given kind of expo curve.

![Screen Shot 2020-02-19 at 09 53 32](https://user-images.githubusercontent.com/11737748/74785355-428e1500-52fe-11ea-9b1b-75af606f1868.jpg)

ctzsnooze commented:
I am not good at spreadsheets, but this one allows visualisation of old vs new rate paradigms.  

https://www.dropbox.com/s/t2spojipp9msuqv/RatesDev.xlsx?dl=0

ctzsnooze commented:
Finally, I want to say how easy it is to make adjustments the new way.  
As a kind of experiment, I just wanted to make a centre sensitivity of 135 and max rate of 500, with a flattish sort of expo curve, for racing.
With the new system, it was really simple, I just set '135' and '500', directly, then tweaked the expo, until the rate curve looked 'right' , which in this case was 0.45.
With the old system, the numbers turned out to be weird, eg 0.67 and 0.73, expo 0, which don't make any intuitive sense.  I couldn't then make the expo curve I wanted.  Even with no 'expo', I couldn't make it flat enough.  

![Screen Shot 2020-02-19 at 10 23 41](https://user-images.githubusercontent.com/11737748/74786897-ecbb6c00-5301-11ea-870a-0442201d844b.jpg)

So I increased my target expo to 0.8 in the new system, which should be achievable by the old system.  The numbers were 135, 500, 08.  I then tried to approximate this steeper expo curve using the old system, with the intent of maintaining the same centre stick feel and max rate.  I found myself chasing my tail.  

It was so annoying!  If I changed the old expo value, it changed centre sensitivity.  If I changed SRate, it changed max roll rate.  Just trying to alter expo alone, leaving centre and max rates where I wanted them,  was impossible.  It was an exercise in frustration.   I in the end, I got close, but the values required were totally, completely different from what I had before - all three of them!  The numbers were 2.3, 0.1 and expo of 0.72.  Totally counter-intuitive.

![Screen Shot 2020-02-19 at 10 31 03](https://user-images.githubusercontent.com/11737748/74787468-60aa4400-5303-11ea-81d5-282be542a613.jpg)

The new approach made the same task really, really easy.  

For the ordinary user, they just set the values, tweak the expo, done.  Tweaking the expo is simple and quick.

etracer65 commented:
While I don’t necessarily have a better one, “ACTUAL” is not a good name for the rates.

JyeSmith commented:
I like it.  easy to remember and communicate to some one else.

And thanks for always doing a great write up with your PRs!

"LEGIBLE" rates???  :smirk:

ctzsnooze commented:
I thought it was less of a mouthful than `CTZSNOOZE` rates lol  :-)
I'm happy for whatever seems to be the most suitable name.

IllusionFpv commented:
I'm not very happy with this (I have some name suggestions but I'll not write here)

mikeller commented:
Interesting work, @ctzsnooze. And a good way to introduce this as an alternative.
Not sure if it is realistic to expect this to become a replacement for the existing rates models, existing users will want to stay with the rates models they are familiar with. So changing anything to 16 bit values will not be that simple. I also have my doubts that 16 bit resolution is required for any of this.
As a way to get around this, and allow users to use intutitive numbers when setting this up, I think we should consider adding MSP support for the rates model selection, and use the information to change the scale for the rate input elements in configurator based on model selection - the currently used fractions are not much good for anything except users of the legacy Betaflight rates model knowing them well.

klutvott123 commented:
Nice. I was experimenting with something similar a while ago. 
Using your code I'm able to create something similar to my rates( 2.23 RC Rate, 0 sRate, 0.7 expo) but just barely. RC Rate has to go all the way up to 255. Could RC Rate have 10 degree resolution like sRate maybe? Is 1 degree resolution really needed?

ctzsnooze commented:
@klutvott123 - thanks for trying this out and your positive feedback. 

 I too had concerns about whether 255 for the centre rate would be enough for everyone.  I thought most likely it would be alright, and I'm glad it was enough in your case.  I use a 1.1 rcRate, with no expo, leading to a 220 centre rate, but I'm 'unusually high' according to those I fly with, and I haven't yet seen a log where the pilot was flying over 255.  But maybe there are people who want more than 255.

We have a number of options:
- We could say, "This method only supports up to 255 deg/s centre sensitivity, if you want more, use the classic Betaflight way of configuring rates."  Not very friendly :-) However we could do it this way and find out if many people were unhappy.
- We could use 10 deg/s resolution, so that a CLI value of 10 means 100 deg/s, and we can only adjust in 10deg/s steps.  So I would enter 0.22 in the Configurator to get my 220 deg/s centre rate.  Someone at 80 would find the next step was 90.  I honestly don't know if that is enough resolution.  In practice, the vast majority of people would use a value under 25 of the 255 available steps, so this seems a bit like we aren't using the potential resolution effectively.  
- Keep the rcRate number the same as it is now, and interpret a 1.0 rcRate value as meaning 200 deg/s centre sensitivity.  In this case, if an MSP update allowed Configurator to show 200 as input in the GUI, we would have the confusing situation of seeing 100 in the CLI.  This I think would be unduly confusing.

I think I'd like to leave it 1 deg/s resolution and see if anyone finds that inadequate, in practice.  I will ask around in the meantime.  If the 255 limit is a problem, then likely we would need to reduce resolution to 10deg/s.

Please ask around for rates settings used by people who say that they run steep centre rates?

klutvott123 commented:
@ctzsnooze I thought my rates were low but apparently not. Maybe I have them higher than "normal" because I'm using no or low(40-50) feedforward. Around center they are about 320/330 degrees/second.
If I were to drop my expo any lower than 0.7 I would have a real problem recreating my rates with ACTUAL rates. Also I wouldn't be able to use ACTUAL rates rcExpo below 0.55.

Since we're limited to 8 bits for rcRate neither 1 or 10 degree/second resolution is the perfect solution, but I think more people would be able to use this with 10 degree/second resolution. 

Keeping rcRate as is would just defeat the purpose of this PR(numbers that actually make sense :smile: ).

bexamous commented:
Is excel spreadsheet correct?  Eg put in old rates 1.75/0.80/0.0.  Okay so then new rates would be ~0.35/1.75/0.70.  I kinda like linear-as-possible near center and then only further out becomes non-linear and ramps up...  I mean usually I end up doing stuff on Taranis, eg: https://i.imgur.com/KYqWORS.jpg

But playing with new rates, at least using excel spreadsheet above...  ~0.35/1.75/0.70 .. okay what if I want sharper bend to keep it more liner near center... 0.30/1.75/0.95 .. or 0.30/1.75/1.0 .. or 0.30/1.75/1.2  oops, don't want that... pretty sure no one would ever want that.

ctzsnooze commented:
@bexamous - excellent, a set of LOS rates :-)

klutvott123 commented:
@ctzsnooze I think you need to use ``rcCommandfAbs`` instead of ``rcCommandf`` in here: https://github.com/betaflight/betaflight/blob/3bcdc33d93ee82fdeb6a740839f67603b9fb7733/src/main/fc/rc.c#L199

This is the result with the current code:
![rates](https://user-images.githubusercontent.com/41271048/75151411-e6b0fb00-5706-11ea-9dd5-2daba0180f5c.png)

@bexamous I like your solution too. It's simple and has full range of adjustment. Allows for all sorts of silly rates. lol

KarateBrot commented:
You could call it something like "normalized rates" since the maximum rate is fixed or normalized to a desired value.

ctzsnooze commented:
Thanks @KarateBrot I pushed a 'fix' to an error that wasn't real.  Have restored the original code and it now works properly.  Sorry for the diversion.

IllusionFpv commented:
Hi guys, you can try and support the original idea -> https://github.com/betaflight/betaflight/pull/9506

erikspen commented:
This sounds eerily similar to the "Alt Tuning Params" I implemented here several months ago: https://erikspen.github.io/betaflightratestuner

KarateBrot commented:
> 
> 
> This sounds eerily similar to the "Alt Tuning Params" I implemented here several months ago: https://erikspen.github.io/betaflightratestuner

Something seems off, because when I set SuperRate and Expo both to 0, the resulting line is not linear in your graph.

ctzsnooze commented:
We now have an  at [https://kmitchel.github.io]('Actual Rates' simulator) thanks to iCr4sh.  It came from the original code by apocolipse for RotorPirates.  We found their code had an error in how it calculated expo, that error has been corrected in this simulator.

@erikspen - My apologies, I was unaware of your work.  Yes, the principles are exactly the same.  There are some differences between the range of expo available in this PR.

ctzsnooze commented:
@klutvott123 I hope I have addressed your concerns about limited centre rate sensitivity by changing the scaling factor.  Here are examples of 350deg centre sensitivity in Actual Rates, rcRate = 0.35, with max rate at 800, and either 0 expo or 1 of expo.  There is a good range of expo available to you compared to defaults (more or less is possible) while keeping 350 deg/s in centre.
Maybe this will work OK for you now?
![Screen Shot 2020-02-27 at 19 24 50](https://user-images.githubusercontent.com/11737748/75425634-34dc2d80-5997-11ea-9fe5-b7b947bcb327.jpg)

KarateBrot commented:
I really like how you can modulate the slope of the curves. So many possibilities to try...

erikspen commented:
@ctzsnooze  Np.

One thing I noticed when comparing to FlightOne rates is their equivalent to Super rate actually grows a bit faster in the middle of the rate curve.   This explains why high-expo freestyle rates in Betaflight are hard to duplicate (exactly) in FlightOne.  So, like a "growth factor" of Super rate would be interesting (how soon the slope really starts to ramp up near the end), ...but that end up adding a 4th parameter to an already complicated thing, ...not worth it.

erikspen commented:
Finally had a good look and really like what you've proposed here.

- Love how the new expo acts over such a large range, can really tweak the shape of the curve. If you play with the "Shape" parameter on my site (https://erikspen.github.io/betaflightratestuner) you'll notice that the old parameters didn't provide too much room for adjustment, though it was "enough" for most people.
- Also like how the new Super is roughly Max deg/s / 1000.  That's pretty intuitive.
- Like how the new Rc Rate doesn't have the odd 200x factor applied, just a straight up 100x to understand how may deg/s it equates to (if linear).  People can just divide their old Rc Rate by half as a starting point.

KarateBrot commented:
When trying to shape the new curve exactly like the corresponding BF curve, it's very easy to recreate the shape. But the other way around if you try to get a BF curve shaped exactly like the new curve, it's very difficult or even impossible. 

So one could argue that the new rates are superior to the BF rates because there are **more possibilities for the shape of the curve**.

Ozzy33 commented:
When I enter my BF rates (2.50, 1950d/s) and rc-expo (0.4) I get the curve, but then to see what matches for the new proposed rate system, only a linear line shows for actual rates (blue line), super and rc expo seems to do nothing.
![BF new rates calc](https://user-images.githubusercontent.com/10664034/75516868-39faa600-59b2-11ea-9598-a7c1ae69def3.jpg)

KarateBrot commented:
You need to keep Center sensitivity smaller than Max Rate.

Asizon commented:
@Ozzy33 for 195d/s you need to put 0,19

Ozzy33 commented:
Thanks for the tips (the super expo threw me off). Even setting max rate to max, i only get 1000d/s maximum, how do i match my 1950d/s and 0.4 rc-expo with the new system? tia

klutvott123 commented:
@Ozzy33 "Max Rate" is limited to 1 in the calculator, but it should allow full range 0-2,55 allowing up to 2550 d/s.
![ozzy33rates](https://user-images.githubusercontent.com/41271048/75544069-7fa38700-5a23-11ea-8a84-93bc6ffe9521.png)

@ctzsnooze Looks good! I'm able to make a reasonable approximation of my rates. It's a little different, but would probably work just fine. I haven't actually flown it . Too cold. Also threw in the @IllusionFpv quick rates for comparison.
![fdhjgdgfj](https://user-images.githubusercontent.com/41271048/75566035-57ca1880-5a4f-11ea-8560-98f5989e97d9.png)

IllusionFpv commented:
@klutvott123 in betaflight we can have max 2000d/s. If you have my spreadsheet, can you try changing expo power for curve(from ^3 to ^2) thanks for testing

Ozzy33 commented:
@ klutvott123
"Max Rate" is limited to 1 in the calculator, but it should allow full range 0-2,55 allowing up to 2550 d/s."  But it really limits it to 1000d/s as seen in the left picture.

To get the max roll rate up to 1950, i needed to set as right picture settings, and then expo does not function. Seems broken :)
![image](https://user-images.githubusercontent.com/10664034/75580336-6ea34780-5a1c-11ea-8380-fcc98996862d.png)
It only takes a second to try it https://kmitchel.github.io/
set BF to 2.50, 0, 0.4
then see if you can match the curve with the new actual settings.

Also, visually what is slightly messing with me is the 2:1 aspect ratio the graph is displayed in (width to height), this makes it look like much more expo in the graph than there is, I think it should be 1:1 and a linear (no expo) line plots at 45 degree angle.

klutvott123 commented:
@IllusionFpv Then the range should be 0-2 for "actual rates". I will try changing the exponent and compare.

@Ozzy33 Yes, super rate = 1 corresponds to 1000d/s for "actual rates". It's just using the same limit as for the regular betaflight rates, but it should be raised to 2000d/s. Then to get your rates you would use the numbers from the first graph above.
A little off topic but I just did some LOS in velocidrone using your rates. Surprisingly I was able to fly it. Hardest part was doing just ONE flip. Easier to just do them in multiples of 10 or so lol.

@ctzsnooze I hope we won't be stuck with a 1000d/s limit 😄

Ozzy33 commented:
@ klutvott123
Thanks for the feedback. A single flip on my quad is 230~250ms, with shortest sticks on my Taranis, it's simply snapping the stick to max and back is one roll. My quad is now old school tech (my frame is 4yo, but motor/esc are updated to 2.5yo tech, and its a tanker, but with the rates high, its snappy).
When I drop to my old rate of 1750, it feels like i am flying a 3s (vs 4s) battery, and when i drop to my older rate of 1200 (quadmovers rates too), it feels like I am pouring molasses in alaska winter or in the movie inception. :)

At the time the DJI guys where seeing how high they could go, I made a joke of how low i could go (and roll). It was interesting to see how the quad moves to the right on a right roll.
[https://www.youtube.com/watch?v=uku7cEiqVIo&t=](url)

My record continuous rolls is 80.

Ozzy33 commented:
Has the 1000d/s limit been addressed (at least seen in the two calculators)? tia

klutvott123 commented:
@Ozzy33 You should be able to set higher values in CLI for now. 

I've seen your videos multiple times already. Impressive flying!

> it feels like I am pouring molasses in alaska winter

Exactly how my rates feel like after flying yours lol. 😄

Ozzy33 commented:
@klutvott123 
I wanted to see what the new numbers are in the graphs that match my current rates/expo, I dont mind punching them in cli. What I did not want to do is constantly experiment to figure out what i want. If BF lets us just choose the the old or new  method I am good, then when the BFC gets updated, then i will try the new rates method.

With higher rates, the quad really feels snappy, this allows you to toss in rolls at places that are not common (near the ground, end of another trick, etc). I been flying the higher rates for ~3 years, before PID FF came around, when I tried FF, it really didn't seem to help. Thanks for your comments.

mikeller commented:
@Ozzy33: We are planning to make selection of the rates model available in the UI in Betaflight configurator. Not sure if we'll be able to add full support for rates curves and custom limits in configurator before 4.2 is due, but it is definitely the plan to get to this at some point.

ugpfpv commented:
Just compiled a clone of Emuflight with your 'actual rates' code, I have been flying the old raceflight rates mode as I didn't care for the stock BF mode, well I can tell you this is my new goto rates mode, very easy to get close to my RF rates only like them even better. Seeing as it's been a while since I've used BF... does the configurator gui show the proper rate curve for Actual Rates? Thanks.

KarateBrot commented:
@ugpfpv It will do in the upcoming configurator 10.7

fpvdoc commented:
Please let me know if this isn‘t the right place for a feature request. I would like to be able to change the rc rate in single digit numbers in order to adjust even more versatile.
